Add number of features used at each step to RFECV.cv_results_

miguelcsilva opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Addresses the suggestion made [here](https://github.com/scikit-learn/scikit-learn/issues/28580#issuecomment-2007439553).


#### What does this implement/fix? Explain your changes.
Add a new key to the `RFECV.cv_results_` dictionary. This key is named `n_features` and its value is a numpy array with the number of features used at each step of the recursive feature elimination process.
It also adds a new test that verifies: 1) the added array is correct; 2) the size of all arrays of this dict is the same.
Finally, it updates the docs [here](https://scikit-learn.org/stable/auto_examples/feature_selection/plot_rfe_with_cross_validation.html) to make use of the simplified way to build the plot. See below plot for the rendered version of the new doc page:
![image](https://github.com/scikit-learn/scikit-learn/assets/66789107/e3626c7a-8dbc-4b19-98b1-60e8d168a4f1)


#### Any other comments?
Tried to make the code roughly aligned with current codebase logic, though I'm not sure I've fully been able to adhere to the repo spirit (specially on the type hints in the tests). So feel free to propose any changes/corrections.

jeremiedbb commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -566,21 +566,20 @@ class RFECV(RFE):
         The fitted estimator used to select features.
 
     cv_results_ : dict of ndarrays
-        All arrays (values of the dictionary) are sorted in ascending order
-        by the number of features used (i.e., the first element of the array
-        represents the models that used the least number of features, while the
-        last element represents the models that used all available features).
```
I think we can keep this, it's still useful

jeremiedbb commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -786,11 +785,18 @@ def fit(self, X, y, groups=None):
-        for i in range(scores.shape[0]):
-            self.cv_results_[f"split{i}_test_score"] = scores_rev[i]
+        self.cv_results_ = {
+            "n_features": np.flip(
+                np.append(
+                    np.arange(
+                        start=n_features, stop=self.min_features_to_select, step=-step
+                    ),
+                    self.min_features_to_select,
+                )
+            ),
```
I'd rather add this key last, in case some users rely on the order (they shouldn't but we don't want to break existing code).

jeremiedbb commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -548,6 +548,44 @@ def test_rfecv_std_and_mean(global_random_seed):
+    ],
+)
+def test_rfecv_cv_results_n_features(
+    min_features_to_select: int,
+    n_features: int,
+    step: int,
+    cv_results_n_features: np.typing.NDArray[np.int_],
```
We have decided to not use type hints yet

jeremiedbb reviewed (commented):
Thanks for the PR @miguelcsilva. 

I think it would make things even simpler if `RFE` exposed the list of n_features kept at each iteration as a fitted attribute array. This way you wouldn't have to recompute the list with this complicated logic when building the `cv_results_` dict. It would also make the computation of the final number of features to select easier. The following https://github.com/scikit-learn/scikit-learn/blob/02b7d4409bb8c3f1af7e19c2e9ac60733eb6e4b9/sklearn/feature_selection/_rfe.py#L761-L767

could be replaced by
```py
        scores_sum = np.sum(scores, axis=0)
        n_features_to_select = rfe.n_features_list_[np.argmax(scores_sum)]
```
where `rfe.n_features_list_` is the new attribute, better name to be found.

ogrisel commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -537,7 +537,7 @@ def test_rfecv_std_and_mean(global_random_seed):
     rfecv = RFECV(estimator=SVC(kernel="linear"))
     rfecv.fit(X, y)
-    n_split_keys = len(rfecv.cv_results_) - 2
+    n_split_keys = len(rfecv.cv_results_) - 3
```
Please add an inline comment to explain where the "3" comes from.

jeremiedbb replied:
Instead I changed it to directly select keys starting with "split" which is easier to understand.

ogrisel commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -786,11 +785,18 @@ def fit(self, X, y, groups=None):
+                    ),
+                    self.min_features_to_select,
+                )
+            ),
```
Instead of re-implementing this decreasing stepping logic, I wonder if the `_rfe_single_fit` function could not be changed to return the array of `n_features` that match the score array it returns. This way we would be more confident that those two regions of the code-based cannot diverge in case of refactoring.

Or alternatively, factorize the call to:

```python
np.arange(
    start=n_features, stop=self.min_features_to_select, step=-step
)
```
in a private helper function that is used both in the `RFE` class and here.

jeremiedbb replied:
I made a similar comment https://github.com/scikit-learn/scikit-learn/pull/28670#pullrequestreview-1951457206, proposing to make the list of n_features directly a public attribute of `RFE`, which I think would make things even simpler that what you propose. What do you think @ogrisel ?

ogrisel reviewed (commented):
Thanks very much for the clean PR. In addition to @jeremiedbb's comment, here are a few more:

miguelcsilva committed:
Add n_features to RFECV.cv_results_

miguelcsilva commented:
@ogrisel and @jeremiedbb think I have addressed most of your suggestions.

Did as @jeremiedbb proposed and added a `n_features_fitted_` attribute to the `RFE` class. Couldn't however get this https://github.com/scikit-learn/scikit-learn/pull/28670#pullrequestreview-1951457206 suggestion to work. The problem is that in case of a tie for best score, the one with the lowest number of features must be selected (there is a test validating that).

There are still two tests failing that I'm not quite sure how to get passing. The first is that on the below, when `.set_params(n_jobs=2)` is called the `rfecv` object loses the `n_features_fitted_` attribute that was set before
https://github.com/scikit-learn/scikit-learn/blob/9b7d176b5a85d7dd681ffa69e55a82a3338096c5/sklearn/feature_selection/tests/test_rfe.py#L432

The second error is on `test_fit_docstring_attributes`. I'm not quite sure how to work with the doctsrings in this repo yet.

Finally, it seems like I also have an error on the changelog even though I added the changes to the changelog for v1.5. Also not sure if I'm doing something wrong here - should have added to v1.4 instead?

Would be great if one of the maintainers could steer me in the right direction on the points above. Hopefully I could then learn this for next time.

ogrisel commented:
Could you please add an entry to the changelog under `doc/whats_new/v1.5.rst`?

miguelcsilva committed:
Fix pr in changelog

jeremiedbb committed:
make new attr non public + simplify argmax

jeremiedbb committed:
Merge branch 'add-n_features-rfecv-cv_results_' of https://github.com/miguelcsilva/scikit-learn into pr/miguelcsilva/28670

jeremiedbb commented:
Thanks @miguelcsilva. I directly pushedsome changes to fix the CI issues.

> Did as @jeremiedbb proposed and added a n_features_fitted_ attribute to the RFE class. Couldn't however get this https://github.com/scikit-learn/scikit-learn/pull/28670#pullrequestreview-1951457206 suggestion to work. The problem is that in case of a tie for best score, the one with the lowest number of features must be selected (there is a test validating that).

It could be fixed by reversing the arrays so that in case of tie breaks the lowest number of features is selected. I pushed it back in my last commit. I still find it a lot more readable than before. And I added a comment to explain why we reverse.

> There are still two tests failing that I'm not quite sure how to get passing. The first is that on the below, when .set_params(n_jobs=2) is called the rfecv object loses the n_features_fitted_ attribute that was set before

Actually I reverted making it a public attribute. I only made it a fitted attribute only accessible by RFECV, like the `scores_` attributes, which makes things easier and keep scope of the PR more focused. This CI issue is thus no longer relevant.

jeremiedbb reviewed (approved):
LGTM

jeremiedbb commented on doc/whats_new/v1.5.rst:
```diff
@@ -188,6 +188,11 @@ Changelog
   :pr:`28085` by :user:`Neto Menoci <netomenoci>` and
   :user:`Florin Andrei <FlorinAndrei>`.
 
+- |Enhancement| The `cv_results_` attribute of :class:`feature_selection.RFECV`has
+  a new key, `n_features`, containing an array with the number of features selected
```
```suggestion
- |Enhancement| The `cv_results_` attribute of :class:`feature_selection.RFECV` has
  a new key, `n_features`, containing an array with the number of features selected
```

jeremiedbb committed:
cln

jeremiedbb committed:
add meson-python to requirements

ogrisel commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -296,6 +299,7 @@ def _fit(self, X, y, step_score=None, **fit_params):
         ranking_ = np.ones(n_features, dtype=int)
 
         if step_score:
+            self.n_features_selected_ = []
```
I find this name a bit misleading: by reading just the name I would expect a final integer at the end of the RFE process.

Maybe `n_features_at_step_` or `n_features_per_iter_` would be more intuitive?

ogrisel replied:
Or we could rename both attributes (they are private, right?) to something like: `step_scores_` and `step_n_features_`.

miguelcsilva replied:
Like that better as well. Just pushed those changes in the latest commit.

ogrisel reviewed (approved):
Nitpicking about variable names, but LGTM, whatever your final decision.

miguelcsilva committed:
Rename private vars

miguelcsilva commented:
> Thanks @miguelcsilva. I directly pushedsome changes to fix the CI issues.
> 
> > Did as @jeremiedbb proposed and added a n_features_fitted_ attribute to the RFE class. Couldn't however get this [#28670 (review)](https://github.com/scikit-learn/scikit-learn/pull/28670#pullrequestreview-1951457206) suggestion to work. The problem is that in case of a tie for best score, the one with the lowest number of features must be selected (there is a test validating that).
> 
> It could be fixed by reversing the arrays so that in case of tie breaks the lowest number of features is selected. I pushed it back in my last commit. I still find it a lot more readable than before. And I added a comment to explain why we reverse.
> 
> > There are still two tests failing that I'm not quite sure how to get passing. The first is that on the below, when .set_params(n_jobs=2) is called the rfecv object loses the n_features_fitted_ attribute that was set before
> 
> Actually I reverted making it a public attribute. I only made it a fitted attribute only accessible by RFECV, like the `scores_` attributes, which makes things easier and keep scope of the PR more focused. It was also needed to be returned by `fit_single_rfe` because when fitting in parallel, the fitted rfe are in subprocesses and not available from the main process. Hence the missing attribute error you got.
> 
> > The second error is on test_fit_docstring_attributes. I'm not quite sure how to work with the doctsrings in this repo yet.
> 
> Since I made the attribute not public, this CI issue is thus no longer relevant.

Thanks for your help with this @jeremiedbb. Just read through the changes you made. Very insightful. Will try to keep it in mind for the next one.

jeremiedbb committed:
a little bit more renaming

jeremiedbb committed:
Merge branch 'add-n_features-rfecv-cv_results_' of https://github.com/miguelcsilva/scikit-learn into pr/miguelcsilva/28670

jeremiedbb committed:
oops

jeremiedbb reviewed (approved):
LGTM. Thanks @miguelcsilva !
